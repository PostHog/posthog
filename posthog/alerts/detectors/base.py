"""Base classes and types for alert detectors."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional


class DetectorType(StrEnum):
    """Available detector types."""

    THRESHOLD = "threshold"
    ZSCORE = "zscore"
    MAD = "mad"  # Median Absolute Deviation


class DetectionDirection(StrEnum):
    """Direction for anomaly detection."""

    UP = "up"  # Alert on values above normal
    DOWN = "down"  # Alert on values below normal
    BOTH = "both"  # Alert on values above or below normal


class ValueType(StrEnum):
    """Whether to run detection on raw values or deltas."""

    RAW = "raw"  # Run detection on raw metric values
    DELTA = "delta"  # Run detection on period-over-period changes


@dataclass
class DetectionResult:
    """Result of running a detector on data."""

    value: Optional[float]  # The raw metric value
    detector_score: Optional[float]  # The detector-specific score (e.g., z-score, mad score)
    is_breach: bool  # Whether this constitutes a breach
    breach_messages: list[str]  # Human-readable breach descriptions
    metadata: dict[str, Any]  # Additional detector-specific info


class BaseDetector(ABC):
    """Base class for all alert detectors."""

    def __init__(self, config: dict[str, Any]):
        """Initialize detector with configuration."""
        self.config = config
        self.validate_config()

    @abstractmethod
    def detect(
        self, values: list[float], series_name: str = "Series", value_type: ValueType = ValueType.RAW
    ) -> DetectionResult:
        """
        Run detection on a list of values.

        Args:
            values: Time series values (most recent last)
            series_name: Name of the series for breach messages
            value_type: Whether to use raw values or deltas

        Returns:
            DetectionResult with breach status and details
        """
        pass

    @abstractmethod
    def validate_config(self) -> None:
        """Validate the detector configuration. Raise ValueError if invalid."""
        pass

    @property
    @abstractmethod
    def detector_type(self) -> DetectorType:
        """Return the type of this detector."""
        pass

    def _compute_deltas(self, values: list[float]) -> list[float]:
        """Compute period-over-period deltas from values."""
        if len(values) < 2:
            return []
        return [values[i] - values[i - 1] for i in range(1, len(values))]
