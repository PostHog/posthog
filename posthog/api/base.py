from rest_framework.pagination import CursorPagination as RestCursorPagination


class CursorPagination(RestCursorPagination):
    ordering = "-created_at"
    page_size = 100
