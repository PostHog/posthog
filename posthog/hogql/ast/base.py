import re
from typing import Optional, Literal

from pydantic import BaseModel, Extra

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException

# NOTE: when you add new AST fields or nodes, add them to the Visitor classes in visitor.py as well!

camel_case_pattern = re.compile(r"(?<!^)(?=[A-Z])")


class AST(BaseModel):
    class Config:
        extra = Extra.forbid

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        from posthog.hogql.ast import UnknownType

        return UnknownType()


class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self


class Expr(AST):
    type: Optional[Type]


class Macro(Expr):
    name: str
    expr: Expr
    # Whether the macro is an inlined column "SELECT 1 AS a" or a subquery "SELECT a AS (SELECT 1)"
    macro_format: Literal["column", "subquery"]
