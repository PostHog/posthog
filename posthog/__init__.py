import sys
from unittest.mock import MagicMock
import deltalake.writer

# DLT imports the missing function `try_get_deltatable` from `deltalake` which
# it doesn't actually use but also no longer exists. So we can safely mock
# it to allow the app to startup without module errors
deltalake.writer.try_get_deltatable = MagicMock()  # type: ignore
sys.modules["deltalake.writer"] = deltalake.writer


# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from posthog.celery import app as celery_app  # noqa: E402

__all__ = ("celery_app",)

# snowflake-connector-python tries to access a root folder which errors out in pods.
# This sets the snowflake home directory to a relative folder
import os  # noqa: E402

os.environ["SNOWFLAKE_HOME"] = "./.snowflake"
