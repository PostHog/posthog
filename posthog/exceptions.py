from typing import Optional, TypedDict

import structlog
from django.conf import settings
from django.http.request import HttpRequest
from django.http.response import JsonResponse
from rest_framework import status
from rest_framework.exceptions import APIException
from sentry_sdk import capture_exception

logger = structlog.get_logger(__name__)


class RequestParsingError(Exception):
    pass


class EnterpriseFeatureException(APIException):
    status_code = status.HTTP_402_PAYMENT_REQUIRED
    default_code = "payment_required"

    def __init__(self, feature: Optional[str] = None) -> None:
        super().__init__(
            detail=(
                f"{feature.capitalize() if feature else 'This feature'} is part of the premium PostHog offering. "
                + (
                    "To use it, subscribe to PostHog Cloud with a generous free tier: https://app.posthog.com/organization/billing"
                    if settings.MULTI_TENANCY
                    else "To use it, get a self-hosted license: https://license.posthog.com"
                )
            )
        )


class EstimatedQueryExecutionTimeTooLong(APIException):
    status_code = 512  # Custom error code
    default_detail = "Estimated query execution time is too long"


class ExceptionContext(TypedDict):
    request: HttpRequest


def exception_reporting(exception: Exception, context: ExceptionContext) -> None:
    """
    Determines which exceptions to report and sends them to Sentry.
    Used through drf-exceptions-hog
    """
    if not isinstance(exception, APIException):
        logger.exception(exception, path=context["request"].path)
        capture_exception(exception)


def generate_exception_response(
    endpoint: str,
    detail: str,
    code: str = "invalid",
    type: str = "validation_error",
    attr: Optional[str] = None,
    status_code: int = status.HTTP_400_BAD_REQUEST,
) -> JsonResponse:
    """
    Generates a friendly JSON error response in line with drf-exceptions-hog for endpoints not under DRF.
    """

    from posthog.internal_metrics import incr

    incr(f"posthog_cloud_raw_endpoint_exception", tags={"endpoint": endpoint, "code": code, "type": type, "attr": attr})
    return JsonResponse({"type": type, "code": code, "detail": detail, "attr": attr}, status=status_code,)
