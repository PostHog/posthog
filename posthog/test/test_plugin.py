import base64

from django.core.exceptions import ValidationError

from posthog.models import Plugin, PluginSourceFile
from posthog.plugins.test.plugin_archives import (
    HELLO_WORLD_PLUGIN_FRONTEND_TSX,
    HELLO_WORLD_PLUGIN_GITHUB_INDEX_JS,
    HELLO_WORLD_PLUGIN_GITHUB_ZIP,
    HELLO_WORLD_PLUGIN_NPM_INDEX_JS,
    HELLO_WORLD_PLUGIN_NPM_TGZ,
    HELLO_WORLD_PLUGIN_PLUGIN_JSON,
    HELLO_WORLD_PLUGIN_PLUGIN_JSON_WITHOUT_MAIN,
    HELLO_WORLD_PLUGIN_RAW_WITH_INDEX_TS_BUT_UNDEFINED_MAIN,
    HELLO_WORLD_PLUGIN_RAW_WITHOUT_ANY_INDEX_TS_AND_UNDEFINED_MAIN,
    HELLO_WORLD_PLUGIN_RAW_WITHOUT_ANY_INDEX_TS_BUT_FRONTEND_TSX,
    HELLO_WORLD_PLUGIN_RAW_WITHOUT_PLUGIN_JS,
)
from posthog.test.base import BaseTest, QueryMatchingTest, snapshot_postgres_queries


class TestPlugin(BaseTest):
    def test_default_config_list(self):
        some_plugin: Plugin = Plugin.objects.create(
            organization=self.organization, config_schema=[{"key": "a", "default": 2}, {"key": "b"}]
        )

        default_config = some_plugin.get_default_config()

        self.assertDictEqual(default_config, {"a": 2})

    def test_default_config_dict(self):
        some_plugin: Plugin = Plugin.objects.create(
            organization=self.organization, config_schema={"x": {"default": "z"}, "y": {"default": None}}
        )

        default_config = some_plugin.get_default_config()

        self.assertDictEqual(default_config, {"x": "z"})


class TestPluginSourceFile(BaseTest, QueryMatchingTest):
    maxDiff = 2000

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_no_archive_fails(self):
        test_plugin: Plugin = Plugin.objects.create(organization=self.organization, name="Contoso")

        with self.assertRaises(ValidationError) as cm:
            PluginSourceFile.objects.sync_from_plugin_archive(test_plugin)

        self.assertEqual(cm.exception.message, f"There is no archive to extract code from in plugin Contoso")

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_zip_without_plugin_js_fails(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization,
            name="Contoso",
            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITHOUT_PLUGIN_JS),
        )

        with self.assertRaises(ValidationError) as cm:
            PluginSourceFile.objects.sync_from_plugin_archive(test_plugin)

        self.assertEqual(cm.exception.message, f"Could not find plugin.json in plugin Contoso")

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_zip_with_explicit_index_js_works(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization, name="Contoso", archive=base64.b64decode(HELLO_WORLD_PLUGIN_GITHUB_ZIP[1])
        )

        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON)
        assert index_ts_file is not None
        self.assertEqual(index_ts_file.source, HELLO_WORLD_PLUGIN_GITHUB_INDEX_JS)
        self.assertIsNone(frontend_tsx_file)

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization, name="Contoso", archive=base64.b64decode(HELLO_WORLD_PLUGIN_NPM_TGZ[1])
        )

        # First time - create
        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON)
        assert index_ts_file is not None
        self.assertEqual(index_ts_file.source, HELLO_WORLD_PLUGIN_NPM_INDEX_JS)
        self.assertIsNone(frontend_tsx_file)

        # Second time - update
        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON)
        assert index_ts_file is not None
        self.assertEqual(index_ts_file.source, HELLO_WORLD_PLUGIN_NPM_INDEX_JS)
        self.assertIsNone(frontend_tsx_file)

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_zip_with_index_ts_works(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization,
            name="Contoso",
            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITH_INDEX_TS_BUT_UNDEFINED_MAIN),
        )

        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON_WITHOUT_MAIN)
        assert index_ts_file is not None
        self.assertEqual(index_ts_file.source, HELLO_WORLD_PLUGIN_GITHUB_INDEX_JS)
        self.assertIsNone(frontend_tsx_file)

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_zip_without_index_ts_but_frontend_tsx_works(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization,
            name="Contoso",
            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITHOUT_ANY_INDEX_TS_BUT_FRONTEND_TSX),
        )

        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON_WITHOUT_MAIN)
        self.assertIsNone(index_ts_file)
        assert frontend_tsx_file is not None
        self.assertEqual(frontend_tsx_file.source, HELLO_WORLD_PLUGIN_FRONTEND_TSX)

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_from_zip_without_any_code_fails(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization,
            name="Contoso",
            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITHOUT_ANY_INDEX_TS_AND_UNDEFINED_MAIN),
        )

        with self.assertRaises(ValidationError) as cm:
            PluginSourceFile.objects.sync_from_plugin_archive(test_plugin)

        self.assertEqual(cm.exception.message, f"Could not find main file index.js or index.ts in plugin Contoso")

    @snapshot_postgres_queries
    def test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works(self):
        test_plugin: Plugin = Plugin.objects.create(
            organization=self.organization,
            name="Contoso",
            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITH_INDEX_TS_BUT_UNDEFINED_MAIN),
        )

        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON_WITHOUT_MAIN)
        assert index_ts_file is not None
        self.assertEqual(index_ts_file.source, HELLO_WORLD_PLUGIN_GITHUB_INDEX_JS)
        self.assertIsNone(frontend_tsx_file)

        test_plugin.archive = base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITHOUT_ANY_INDEX_TS_BUT_FRONTEND_TSX)
        test_plugin.save()

        (plugin_json_file, index_ts_file, frontend_tsx_file,) = PluginSourceFile.objects.sync_from_plugin_archive(
            test_plugin
        )

        self.assertEqual(PluginSourceFile.objects.count(), 2)  # frontend.tsx replaced by index.ts
        self.assertEqual(plugin_json_file.source, HELLO_WORLD_PLUGIN_PLUGIN_JSON_WITHOUT_MAIN)
        self.assertIsNone(index_ts_file)
        assert frontend_tsx_file is not None
        self.assertEqual(frontend_tsx_file.source, HELLO_WORLD_PLUGIN_FRONTEND_TSX)
