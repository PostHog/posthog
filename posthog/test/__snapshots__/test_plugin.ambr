# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.1
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'index.ts'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.2
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('frontend.tsx')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.4
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.5
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'index.ts'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.6
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('frontend.tsx')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_tgz_with_explicit_index_js_works.7
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_explicit_index_js_works
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_explicit_index_js_works.1
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'index.ts'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_explicit_index_js_works.2
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('frontend.tsx')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_explicit_index_js_works.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_index_ts_works
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_index_ts_works.1
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'index.ts'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_index_ts_works.2
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('frontend.tsx')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_with_index_ts_works.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_without_index_ts_but_frontend_tsx_works
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_without_index_ts_but_frontend_tsx_works.1
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'frontend.tsx'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_without_index_ts_but_frontend_tsx_works.2
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('index.ts')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_from_zip_without_index_ts_but_frontend_tsx_works.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.1
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'index.ts'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.2
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('frontend.tsx')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.3
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.4
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'plugin.json'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.5
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" = 'frontend.tsx'
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  LIMIT 21
  FOR
  UPDATE
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.6
  '
  SELECT "posthog_pluginsourcefile"."id",
         "posthog_pluginsourcefile"."plugin_id",
         "posthog_pluginsourcefile"."filename",
         "posthog_pluginsourcefile"."source",
         "posthog_pluginsourcefile"."status",
         "posthog_pluginsourcefile"."transpiled",
         "posthog_pluginsourcefile"."error"
  FROM "posthog_pluginsourcefile"
  WHERE ("posthog_pluginsourcefile"."filename" IN ('index.ts')
         AND "posthog_pluginsourcefile"."plugin_id" = 2)
  '
---
# name: TestPluginSourceFile.test_sync_from_plugin_archive_twice_from_zip_with_index_ts_replaced_by_frontend_tsx_works.7
  '
  SELECT COUNT(*) AS "__count"
  FROM "posthog_pluginsourcefile"
  '
---
